microprofile/connectors/kafka/src/main/java/io/helidon/microprofile/connectors/kafka/BasicKafkaConsumer.java

```diff
@@ -0,0 +1,207 @@
+    // We need to keep the reference here to be able to wake up from pooling when shuting down
+    private final KafkaConsumer<K, V> consumer;
+    private final ScheduledExecutorService scheduler;
+
+    /**
+     * Kafka consumer created from {@link io.helidon.config.Config config}
+     * see configuration {@link KafkaConfigProperties example}.
+     *
+     * @param config Helidon {@link io.helidon.config.Config config}
+     * @param scheduler Helidon {@link java.util.concurrent.ScheduledExecutorService scheduler}
+     */
+    BasicKafkaConsumer(Config config, ScheduledExecutorService scheduler) {
+        this.config = config;
+        this.properties = new KafkaConfigProperties(config);
+        this.topicNameList = properties.getTopicNameList();
+        this.consumer = new KafkaConsumer<>(properties);
+        this.scheduler = scheduler;
+    }
+
+    /**
+     * Create publisher builder.
+     *
+     * @return {@link org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder}
+     */
+    public PublisherBuilder<? extends Message<?>> createPushPublisherBuilder() {
+        return ReactiveStreams.fromPublisher(new BasicPublisher<K, V>(subscriber -> {
+            subscriber.onSubscribe(new Subscription() {
+                @Override
+                public void request(long n) {
+                    // Pushing Kafka consumer doesn't support requests.
+                }
+
+                @Override
+                public void cancel() {
+                    BasicKafkaConsumer.this.close();
+                    LOGGER.log(Level.FINE, "Subscription cancelled.");
+                }
+            });
+            consumer.subscribe(topicNameList, partitionsAssignedLatch);
+            scheduler.scheduleAtFixedRate(new BackPressureLayer(subscriber, 
+                    config.get(POOL_TIMEOUT).asLong().asOptional().orElseGet(() -> 50L)), 0, 
+                    config.get(PERIOD_EXECUTIONS).asLong().asOptional().orElseGet(() -> 100L), TimeUnit.MILLISECONDS);
+        }));
+    }
+
+    /**
+     * Blocks current thread until partitions are assigned,
+     * since when is consumer effectively ready to receive.
+     *
+     * @param timeout the maximum time to wait
+     * @param unit    the time unit of the timeout argument
+     * @throws java.lang.InterruptedException        if the current thread is interrupted while waiting
+     * @throws java.util.concurrent.TimeoutException if the timeout is reached
+     */
+    public void waitForPartitionAssigment(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
+        if (!partitionsAssignedLatch.await(timeout, unit)) {
+            throw new TimeoutException("Timeout for subscription reached");
+        }
+    }
+
+    /**
+     * Close gracefully. Stops wakes possible blocked poll and close consumer.
+     */
+    @Override
+    public void close() {
+        // Stops pooling
+        consumer.wakeup();
+        while (running.get()) {
```

Comment:
I am not sure if this while is acceptable because of CPU usage. However I put it there because the time frame must be very small (few milliseconds) or zero.

Let me know if you prefer other way, with count down latch for example.

Reply:
I will do this with a synchronize.

Reply:
In this case I think it is much better to use a lock.